src/insight/networkInsight.go

```diff
@@ -0,0 +1,269 @@
+		}
+
+		pbEgress.ToPorts = pbToPorts
+		pbEgress.ToCIDRs = pbToCIDRs
+		pbEgress.ToServices = pbToServices
+		pbEgress.ToFQDNs = pbToFQDNs
+		pbEgress.ToHTTPs = pbToHTTPs
+		pbEgresses = append(pbEgresses, &pbEgress)
+	}
+
+	// Spec Ingress
+	for _, ingress := range policy.Spec.Ingress {
+		pbIngress := ipb.Ingress{}
+		pbToPorts := []*ipb.SpecPort{}
+		pbToHTTPs := []*ipb.SpecHTTP{}
+		pbFromCIDRs := []*ipb.SpecCIDR{}
+
+		pbIngress.MatchLabels = ingress.MatchLabels
+		pbIngress.FromEntities = append(pbIngress.FromEntities, ingress.FromEntities...)
+
+		for _, toPort := range ingress.ToPorts {
+			pbToPort := ipb.SpecPort{}
+			pbToPort.Port = toPort.Port
+			pbToPort.Protocol = toPort.Protocol
+			pbToPorts = append(pbToPorts, &pbToPort)
+		}
+
+		for _, toHTTP := range ingress.ToHTTPs {
+			pbToHttp := ipb.SpecHTTP{}
+			pbToHttp.Path = toHTTP.Path
+			pbToHttp.Method = toHTTP.Method
+			pbToHttp.Aggregated = toHTTP.Aggregated
+			pbToHTTPs = append(pbToHTTPs, &pbToHttp)
+		}
+
+		for _, fromCIDR := range ingress.FromCIDRs {
+			pbFromCIDR := ipb.SpecCIDR{}
+			pbFromCIDR.CIDRs = append(pbFromCIDR.CIDRs, fromCIDR.CIDRs...)
+			pbFromCIDR.Except = append(pbFromCIDR.Except, fromCIDR.Except...)
+			pbFromCIDRs = append(pbFromCIDRs, &pbFromCIDR)
+		}
+
+		pbIngress.FromEntities = append(pbIngress.FromEntities, ingress.FromEntities...)
+
+		pbIngress.ToPorts = pbToPorts
+		pbIngress.ToHTTPs = pbToHTTPs
+		pbIngress.FromCIDRs = pbFromCIDRs
+		pbIngressess = append(pbIngressess, &pbIngress)
+	}
+
+	pbNetSpec.Egressess = pbEgresses
+	pbNetSpec.Ingressess = pbIngressess
+
+	return *pbNetSpec
+}
+
+func GetNetInsightData(req types.InsightRequest) ([]ipb.NetworkInsightData, error) {
+
+	var networkData []ipb.NetworkInsightData
+
+	nwpolicies := libs.GetNetworkPolicies(network.CfgDB, req.ClusterName, req.Namespace, "latest", req.Type, req.Rule)
+
+	for _, nwpolicy := range nwpolicies {
+
+		var locRes ipb.NetworkInsightData
+
+		nwPbSpec := populateNwInsightData(nwpolicy)
+
+		locRes.ClusterName = nwpolicy.Metadata["cluster_name"]
+		locRes.Namespace = nwpolicy.Metadata["namespace"]
+		locRes.Type = nwpolicy.Metadata["type"]
+		locRes.Rule = nwpolicy.Metadata["rule"]
+		locRes.Labels = nwPbSpec.Labels
+		nwPbSpec.Labels = ""
+		locRes.NetResource = append(locRes.NetResource, &nwPbSpec)
+
+		if req.Labels != "" {
+			if locRes.Labels == req.Labels {
+				networkData = append(networkData, ipb.NetworkInsightData{
+					ClusterName: locRes.ClusterName,
+					Namespace:   locRes.Namespace,
+					Labels:      locRes.Labels,
+					Type:        locRes.Type,
+					Rule:        locRes.Rule,
+					NetResource: locRes.NetResource,
+				})
+			}
+		} else {
+			networkData = append(networkData, ipb.NetworkInsightData{
+				ClusterName: locRes.ClusterName,
+				Namespace:   locRes.Namespace,
+				Labels:      locRes.Labels,
+				Type:        locRes.Type,
+				Rule:        locRes.Rule,
+				NetResource: locRes.NetResource,
+			})
+		}
```

Comment:
Please simplify this duplicate code. What will be the case when `req.Labels != "" and req.Labels != locRes.Labels`?